src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_BindAsync_Snapshot.generated.txt

```diff
@@ -8,2222 +8,2188 @@
+                    global::Microsoft.AspNetCore.Http.Generators.Tests.MyBindAsyncFromInterfaceRecord myBindAsyncParam_local;
+                    if ((object?)myBindAsyncParam_temp == null)
+                    {
+                        logOrThrowExceptionHelper.RequiredParameterNotProvided("MyBindAsyncFromInterfaceRecord", "myBindAsyncParam", "MyBindAsyncFromInterfaceRecord.BindAsync(HttpContext)");
+                        wasParamCheckFailure = true;
+                        myBindAsyncParam_local = default!;
                     }
-
-                    async Task RequestHandler(HttpContext httpContext)
+                    else
                     {
-                        var wasParamCheckFailure = false;
-                        var myBindAsyncParam_temp = await BindAsync<global::Microsoft.AspNetCore.Http.Generators.Tests.BindAsyncFromExplicitStaticAbstractInterface>(httpContext, parameters[0]);
-                        var myBindAsyncParam_local = (global::Microsoft.AspNetCore.Http.Generators.Tests.BindAsyncFromExplicitStaticAbstractInterface?)myBindAsyncParam_temp;
-
-                        if (wasParamCheckFailure)
-                        {
-                            httpContext.Response.StatusCode = 400;
-                            return;
-                        }
-                        var result = handler(myBindAsyncParam_local);
-                        if (result is string)
-                        {
-                            httpContext.Response.ContentType ??= "text/plain; charset=utf-8";
-                        }
-                        else
-                        {
-                            httpContext.Response.ContentType ??= "application/json; charset=utf-8";
-                        }
-                        await httpContext.Response.WriteAsync(result);
+                        myBindAsyncParam_local = (global::Microsoft.AspNetCore.Http.Generators.Tests.MyBindAsyncFromInterfaceRecord)myBindAsyncParam_temp;
                     }
 
-                    async Task RequestHandlerFiltered(HttpContext httpContext)
+                    if (wasParamCheckFailure)
                     {
-                        var wasParamCheckFailure = false;
-                        var myBindAsyncParam_temp = await BindAsync<global::Microsoft.AspNetCore.Http.Generators.Tests.BindAsyncFromExplicitStaticAbstractInterface>(httpContext, parameters[0]);
-                        var myBindAsyncParam_local = (global::Microsoft.AspNetCore.Http.Generators.Tests.BindAsyncFromExplicitStaticAbstractInterface?)myBindAsyncParam_temp;
-
-                        if (wasParamCheckFailure)
-                        {
-                            httpContext.Response.StatusCode = 400;
-                        }
-                       
```

Comment:
Why doesn't this just happen inline? It may make sense to throw a comment in the generated code...

Reply:
The parameter binding logic that sets the `wasParamCheckFailure`is shared between the non-filter and filter-handler. However, the way we react to it is different depending on whether we are executing an endpoint with filters or not.

When no filters are registered, we return from the RequestDelegate immediately. When filters are registered, we don't return because we want the filters to run, but _not_ the request delegate. This is easier to do if the check doesn't happen inline.

Reply:
I think GitHub is messing up showing where I was referring to. When you switch to the "Files" tab, you can see where I was referring to:

![image](https://github.com/dotnet/aspnetcore/assets/8291187/95e1f123-ee7f-4b1e-9f2b-6e2ae5491bc0)

Why is the `Cast` helper method necessary? Why can't we just cast inline without using this method?

Reply:
> Why is the Cast helper method necessary? Why can't we just cast inline without using this method?

Ah, that part is in reaction to [a breaking change in the compiler](https://github.com/dotnet/csharplang/blob/main/proposals/lambda-method-group-defaults.md#breaking-change). Method group references are now inferred as anonymous delegate types instead of `Action`/`Func` to match the conversion for lambdas. This is the workaround @jjonescz and I came up with given the new constraint.

Reply:
Cast is more like Infer. Using generics to reference and unspeakable type (like an anonymous class).